build_tools/_therock_utils/artifact_backend.py

```diff
@@ -0,0 +1,289 @@
+"""Abstraction layer for artifact storage backends (S3 or local directory).
+
+This module provides a unified interface for artifact storage that works with
+both local directories (for prototyping/testing) and S3 (for CI/CD).
+
+Environment-based switching:
+- THEROCK_LOCAL_STAGING_DIR set → use LocalDirectoryBackend
+- Otherwise → use S3Backend with existing retrieve_bucket_info() logic
+"""
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from pathlib import Path
+from typing import List, Optional, Set
+import os
+import shutil
+
+
+@dataclass
+class ArtifactLocation:
+    """Represents an artifact's location in the backend."""
+
+    artifact_key: str  # e.g., "blas_lib_gfx94X.tar.xz"
+    full_path: str  # Backend-specific full path/URI
+
+
+class ArtifactBackend(ABC):
+    """Abstract base for artifact storage backends."""
+
+    @abstractmethod
+    def list_artifacts(self, name_filter: Optional[str] = None) -> List[str]:
+        """List available artifact filenames.
+
+        Args:
+            name_filter: Optional artifact name prefix to filter by (e.g., "blas" to match "blas_lib_*.tar.xz")
+
+        Returns:
+            List of artifact filenames (e.g., ["blas_lib_gfx94X.tar.xz", "blas_dev_gfx94X.tar.xz"])
+        """
+        pass
+
+    @abstractmethod
+    def download_artifact(self, artifact_key: str, dest_path: Path) -> None:
+        """Download/copy an artifact to a local path.
+
+        Args:
+            artifact_key: The artifact filename (e.g., "blas_lib_gfx94X.tar.xz")
+            dest_path: Local path to write the artifact to
+        """
+        pass
+
+    @abstractmethod
+    def upload_artifact(self, source_path: Path, artifact_key: str) -> None:
+        """Upload/copy a local artifact to the backend.
+
+        Args:
+            source_path: Local path of the artifact to upload
+            artifact_key: The artifact filename to use in the backend
+        """
+        pass
```

Comment:
(future optimization) Some batch APIs may be useful here too, i.e.

```python
@abstractmethod
def download_artifacts(self, artifact_keys: Iterable[str], dest_paths: Iterable[Path) -> None
```

I bet the cloud upload/download APIs offer better throughput for batch requests.